runtime/src/bank/transaction_account_state_info.rs

```diff
@@ -0,0 +1,63 @@
+use {
+    crate::{
+        account_rent_state::{check_rent_state, RentState},
+        bank::Bank,
+    },
+    solana_sdk::{
+        feature_set, message::SanitizedMessage, transaction::Result,
+        transaction_context::TransactionContext,
+    },
+};
+
+pub(crate) struct TransactionAccountStateInfo {
+    rent_state: Option<RentState>, // None: readonly account
+}
+
+impl Bank {
+    pub(crate) fn get_transaction_account_state_info(
+        &self,
+        transaction_context: &TransactionContext,
+        message: &SanitizedMessage,
+    ) -> Vec<TransactionAccountStateInfo> {
+        (0..transaction_context.get_number_of_accounts())
+            .map(|i| {
+                let rent_state = if message.is_writable(i) {
+                    let account = transaction_context.get_account_at_index(i).borrow();
+                    Some(RentState::from_account(
+                        &account,
+                        &self.rent_collector().rent,
+                    ))
+                } else {
+                    None
+                };
```

Comment:
For the assert, this is the code that assumes that writable native program accounts wouldn't make it this far. It'd be nice to have at least a debug assert to document that assumption